File: 2.1 Algorithmic Toolbox/Week 2 Introduction/01_introduction_starter_files/fibonacci_sum_last_digit/fibonacci_sum_last_digit.cpp
#include <iostream>
#include <vector>
using std::vector;

vector<int> pisano(int n){
	long long previous = 0;
	long long current = 1;

	vector<int> cycle;
	cycle.push_back(0);
	cycle.push_back(1);
	int i = 1;

	while ( i == 1){
		long long tmp_previous = previous;
		previous = current;
		current = (tmp_previous + current) % n;
		cycle.push_back(current);
		int m = cycle.size();
		vector<int>::const_iterator first = cycle.begin();
		vector<int>::const_iterator mid = cycle.begin() + int(m/2);
		vector<int>::const_iterator last = cycle.end();
		vector<int> v1(first, mid);
		vector<int> v2(mid, last);
		if (m%2 == 0 && m != 0 && v1 == v2){
			break;
		}
	}
	int clen = cycle.size();
	vector<int>::const_iterator first = cycle.begin();
	vector<int>::const_iterator mid = cycle.begin()+int(clen/2);
	vector<int> pisano(first, mid);
	return pisano;
}

int fibonacci_sum_fast(int n) {
    // the nth fibonacci is F(n+2) - 1
	vector<int> cycle;
	cycle = pisano(10);
	int clen = cycle.size();
	return cycle.at((n+2) % clen) -1;
}

int main() {
    long long n = 0;
    std::cin >> n;
	vector<int> cycle;
	cycle = pisano(10);
	int clen = cycle.size();
	std::cout << fibonacci_sum_fast(n) << std::endl;

	system("pause");
}

File: 2.1 Algorithmic Toolbox/Week 2 Introduction/01_introduction_starter_files/fibonacci_sum_last_digit/fibonacci_sum_last_digit.py
# Uses python3
#import sys

def fibonacci_sum_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum      = 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum += current

    return sum % 10

def fib(n, memo = {0: 0, 1: 1}):
    if n not in memo:
        memo[n] = fib(n-1, memo)+fib(n-2, memo)
    return memo[n]

def pisano(n):
    cycle = []
    i = 0
    while True:
        cycle.append(fib(i) % n)
        m = len(cycle)
        if m % 2 == 0 and m != 0 and cycle[:int(m/2)] == cycle[int(m/2):]:
            break
        i += 1
    return cycle[:int(m/2)]

def fibonacci_sum_fast(n):
    # the sum of the nth fibonnaci is F(n+2)-1
    mod10 = pisano(10)
    clen = len(mod10)
    return mod10[(n+2) % clen] - 1

if __name__ == '__main__':
    #input = sys.stdin.read()
    n = int(input())
    print(fibonacci_sum_fast(n))
